benchmark/scenarios/_rate_measurer.py

```diff
@@ -61,9 +61,9 @@ def update_rate(self):
         Update the current rate and record a new sample.
         """
         self._rate = (
-            (self._received - self._samples[0]) / float(self.sample_size)
+            (self._sent - self._samples[0]) / float(self.sample_size)
         )
-        self._samples.append(self._received)
+        self._samples.append(self._sent)
```

Comment:
The code looks good, but it has no way to handle a large number of error responses.  Errors are different to outstanding calls. There is a fixed limit of outstanding calls we should tolerate. Errors on the other hand can be tolerated relative to how many calls we have sent.

I'd suggest changing as follows:
```
    def update_rate(self):
        """
        Update the current rate and record a new sample.
        """
        # Number of calls sent minus number of errors received is our best
        # indicator of "good calls".  Number of calls received has shown to be
        # too lumpy to generate a smooth rate under heavy load.  While errors
        # received might also be lumpy, they are either a small proportion of
        # total calls, or if a large proportion, we should fail anyway.
        good_count = self._sent - self._error_count
        self._rate = (
            (good_count - self._samples[0]) / float(self.sample_size)
        )
        self._samples.append(good_count)
```

This might make the test `test_scenario_throws_exception_when_rate_drops` relevant again :question: